src/lib.rs

```diff
@@ -71,6 +71,8 @@
     clippy::cast_lossless,
     // Covered by other lints
     clippy::missing_panics_doc, // clippy::unwrap_used
+    // Callbacks with `+ 'static + Send + Sync`, trigger this pretty easily
+    clippy::type_complexity,
```

Comment:
```suggestion
```
I think you should instead define a type alias for the collector like:`type Filter<T> = Arc<dyn Fn(&Arc<T>) -> bool + 'static + Send + Sync>;`, and then just replace the uses of it. The next branch essentially has this, but souped up with GATs and the like.